core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -673,7 +673,19 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * otherwise executes the specified effect.
    */
   final def orElse[R1 <: R, E2, A1 >: A](that: => ZIO[R1, E2, A1]): ZIO[R1, E2, A1] =
-    tryOrElse(that, new ZIO.SucceedFn(() => that))
+    self.foldCauseM(
+      ZIOFn(() => that) { cause =>
+        cause.stripFailures match {
+          case None =>
+            that.catchSomeCause {
+              case c2 if c2.died => ZIO.halt(Cause.die(FiberFailure(Cause.Both(cause, c2))))
```

Comment:
Why does it have `if c2.died` here?

Reply:
@jdegoes as far as I understood, I only need to catch died actions. Should I catch failed as well?

Reply:
If I catch both failed and died actions here, I would make the whole action died. I thought it isn't semantically correct.

Reply:
I could handle failed right hand side with something like this:
```scala
case c2 if c2.failed => ZIO.halt(Cause.Both(Cause.die(FiberFailure(cause)), c2))
```